### ISSUE TITLE
Nesting image layers in groups for navigation

### ISSUE THREAD

**Issue by NHPatterson** [napari/napari#970]
## 🚀 Feature
Following discussion [here](https://forum.image.sc/t/image-array-layer-grouping/33908). It could be beneficial to group image planes from the same image. For instance, group multi-channel images together and allow a collapsible view of the channels. Right now, the system assigns a numerical index to the name of a multi-channel image so that ```My MC image``` becomes ```My MC image(1)```, ```My MC image(2)```, etc. in the side bar. It would also be beneficial here to have naming in the arguments of a nested layers, so that one could have like below where both the modality and the channels can be optionally named. 
```
Modality 1
   - DAPI
   - Cool antibody 1
   - Cool antibody 2

Modality 2
   - DAPI
   - Cool antibody 3
   - Cool antibody 4

```

## Motivation
Improve navigation in large datasets with many image planes. One can imagine a scenario where there is a multi-channel image composed of 10s or 100s of images, perhaps different processing of the original plane. Loading them all as single entries in the sidebar without individual naming could be very cumbersome.

## Alternatives
Not sure.

## Additional context
Here is an image of what that may look like, out of context. Of course, maybe hierarchy can be limited to only so many depths. 
![image](https://user-images.githubusercontent.com/17855764/74354223-a5bcfc00-4d80-11ea-96bf-33fc306f1141.png)

**Comment by kevinyamauchi** [comment]
I like this idea!

If layers are nested, should there be some layer properties that are synced between the layers? For example, it seems like it should be possible to show/hide all of the layers in the group with a single click/command. I could also see it being useful to set opacity for all layers in a group together as well. Or maybe it would be better if the `LayerGroup` (or whatever we call it) can set the visibility/opacity for all child layers, but the child layers cannot set the layer properties of their "siblings". What do you all think?

In terms of API, I think it could make sense to add a `group` kwarg to the `Viewer.add_*` methods. If a `group` name is provided, the layer is added to the group if it exists and if the group does not exist, the group gets created and the layer gets added to it. Thoughts?

**Comment by GenevieveBuckley** [comment]
I think grouped layers will be very useful. I think it's also important to be able to support groups of groups as well (so people can have more complicated hierarchies if that's what they need as well).

**Comment by sofroniewn** [comment]
I really like this idea - here is another example where I am running into this, using some multiview tracking data from @lambdaloop and https://github.com/lambdaloop/anipose where there are 3 cameras, each one contributing an image layer, a points layer, and a vectors layer. It would be nice to be able to group those three layers all together for each camera. I think this matches what is being requested here nicely.

The main tricky thing I can see is the design -, making it easy to navigate, intuitive to understand, and look nice, though theres is definitely lots of prior art here in other tools.

One additional thing I see is that we can then use these "Groups" for features like our grid view and multicanvas mode (see #760 say) where say I want to easily be able to assign an entire group to a different canvas. We might want to have some idea of how that will be done before implementing anything here.

<img width="1200" alt="Screen Shot 2020-02-15 at 7 41 48 PM" src="https://user-images.githubusercontent.com/6531703/74598749-f7d57a00-502b-11ea-886d-cfc589d8ef15.png">

**Comment by jojoelfe** [comment]
Hi,

I work on correlated microscopy and have so far been using a tool from Zeiss called ZEN Connect to overlay images from different microscopes. I have been desperate for an open-source tool to do this job and napari looks super promising.

This feature would be fantastic for this kind of data as one quickly builds a up a hierarchy of images. This is a typical example:

```
Light Microscope
|
|----5x
|    |----Brightfield
|    |----GFP
|
|----10x
|    |----Brightfield
|
|----100x
     |----Feature 1
     |	  |----GFP [Z-stack]
     |	  |----RFP [Z-stack]
     |	  |----Reflection
     |		
     |----Feature 2
          |----GFP [Z-stack]
          |----RFP [Z-stack]
          |----Reflection


Scanning Electron Microscope
|----100x
|    |----SE detector
|
|----500x
|    |----SE detector
|
|----1000x
     |----Feature 1
     |    |----SE detector
     |    |----In-lens detector
     |
     |----Feature 2
          |----SE detectore
          |----In-lens detector
   

Transmission Electron Microscope
|----135x
|----2250x
|----81000x
     |----Feature 1
     |----Feature 2
```
I think it would be great if any node of this tree could have settings that trickle down to all children, such as opacity. It would be awesome if any node could have an associated transformation matrix. Then every data plane could be transformed into the world coordinate system ( #763 ) by chaining the transforms as one climbs the tree. Since many elements of the transformation are shared between children (rotation between microscopes, scale between magnification, translation between features of interest) this would be a very efficient way of aligning a lot of images.

**Comment by sofroniewn** [comment]
That's great that you're excited about napari @jojoelfe. Those suggestions for the layer groups and chaining transforms / passing down properties sound like really good ones. I think we might be able to do something like make a new object - a `LayerGroup` that subclasses the base `Layer` so it has all the properties that are common for all the `Layers` and a list of the layers / layergroups it contains and methods that allow it to do the propagation.

Probably worth waiting till  #885 goes in, but then is something we should definitely work on

**Comment by GenevieveBuckley** [comment]
# Summary of meeting discussion

I've summarized our discussion from the meeting today about layergroups.

There are two halves to the problem of how we handle layergroup properties:
1. Getting attributes
2. Setting attributes

## 1. Getting attributes
* This is a little philosophically tricky, because we need to work out what a LayerGroup property should be. 
For example `.opacity` (float), `.visible` (boolean), and `.data` (array) are all layer properties. What should 
* It was initially suggested that maybe we disallow getting attributes from layergroups, but this breaks the composite design pattern & would likely cause other issues down the track. So we won't do that.
* @royerloic suggests we treat layergroup properties as another node on a tree of transforms. Any time you want to retrieve a property, you need to run through all the parents to that node. 
* It's important we allow an "undetermined" state. What's the visibility of a layergroup where parts of the tree are visible and parts are not visible?
* There are two things (sometimes at odds) that we need to support here:
    * Allowing properties to have relative scaling. Example: a multichannel image where each channel looks best with a different opacity, but we also want to use the opacity slider for the whole layergroup to change all the opacities at once (without losing the relative proportions between them).
    * Allowing properties to set absolute values, no relative scaling. Example: we have many images but want to make sure they are all using the same contrast/brightness display settings.

## 2. Setting attributes
* The goal is for LayerGroups to propagate their attributes/properties to children without a lot of code duplication
* @kne42 suggests we use subclass `__init__` instead of metaclasses to achieve this

Next steps:
* Try experimenting with subclass `__init__`
* Try experimenting with Loic's tree of transforms idea

@royerloic, @kne42, @jni, @tlambert03, please add anything you think I've forgotten here.

**Comment by GenevieveBuckley** [comment]
Notes from the meeting:
Attributes on layergroup nodes should *not* propagate downwards, all nodes & leaves should get/set attributes as normal. But, there needs to be a hidden "effective attribute" method that is calculated purely for display/rendering.

**Comment by GenevieveBuckley** [comment]
Ok, here is a thought experiment I need people to weigh in on. 

This is the most succinct way I can describe the larger problem, by using an example about a specific attribute. Last time I brought this up in the group meeting we shelved that discussion, but it really does need an answer. 

**What should the `.data` property of a layergroup be?**

* Should it be None?
* Should it be a list of all child .data results? A generator?
* I can't really stack all the child .data results along a new array axis, because they all have very different shapes. Also, this is much more work to refresh/update on data changes for any of the layers (this criticism also applies to the list/generator idea above, too)
* Should the .data attribute just not exist at all for layergroups?

None of these possible answers seem any good. Juan wants a composite design pattern, where a layergroup can be handled in an identical manner as any single layer.

The worry is that lots of things will break if unexpected things are returned from properties (or if an expected property doesn't exist). Right now, we don't have a clear answer to how much stuff breaks either.

**Comment by GenevieveBuckley** [comment]
To summarize the different aspects that need covering, layergroups have:
1. Properties/attributes related to the positioning of the layer(group) eg: `scale`, `translate`, transform chains, etc., or related to the vispy canvas display. Eg: `visible`, `opacity`, plus properties specific to certain types of layers.
    * Our current approach is that it makes sense for these properties to exist independently on the layergroup. At runtime these properties can be calculated, which allows parent layergroups to modify and return some "effective" value that is actually used by Qt/Vispy.
2. Properties/attributes that may not necessarily make sense. Eg: `data`(which is  
    * We don't have a good answer for this yet.
3. Methods that don't return anything. Eg: refresh(), etc.
    * This is easy: we should probably just iterate through the whole layergroup tree and 
4. Methods that do return something. Eg: get_value(), etc.
    * This is the same problem as in (2), we have no good answer for this yet.

**Comment by tlambert03** [comment]
interesting / helpful points @GenevieveBuckley!

> None of these possible answers seem any good. Juan wants a composite design pattern, where a layergroup can be handled in an identical manner as any single layer.

I was thinking we don't need the composite to have _every_ characteristic of single layers, just the ones that make sense and that we need for the layerlist tree.  I do think it's going to a specific decision for each property... I agree that `.data` can't be combined really, (I like the generator idea the most I think).

I think my main question would be: If you just punt on the layer group `data` attribute issue, at what line exactly in the napari codebase does it become a problem, and why?  I think the answer  to that might help to inform what (if anything) we need from a composite data attribute.

Our current LayerList has _no_ methods that get a value for every layer in the list.  So we're moving in to new territory here.  But that also means that there really isn't anywhere in the codebase yet that demands it.  So if there are properties where it's easy to figure out (scalars or bools like opacity, visibility) great... but I think each method is going to take some thought and there may not be universally applicable strategies.  For instance, maybe `get_extent` gives you a single outer bounding box of of all the other ones, whereas other methods might just iterate over the children?

**Comment by GenevieveBuckley** [comment]
Thanks for your thoughts @tlambert03!

> I think my main question would be: If you just punt on the layer group `data` attribute issue, at what line exactly in the napari codebase does it become a problem, and why? I think the answer to that might help to inform what (if anything) we need from a composite data attribute.

The `.data` property setter is marked as an abstract method on the base layer, so it requires some kind of implementation (even if that implementation is just "raise NotImplementedError"). The other base layer abstract methods are: `_get_extent()`, `_get_ndim()`, `_get_state()`, `_set_view_slice()`, `_update_thumbnail()`.

It's very difficult to tell at which line in the napari codebase these things become problems, because the problems tend to crop up in the emitter event interactions between Qt and Vispy with the python model. The full scope of these interactions isn't clear to me, and I can't continue trying "poke it and see what breaks" without building in more stuff involving Qt/Vispy. That approach was discouraged at our last meeting (it was preferred that we get clear on the python model first), but it's starting to feel like a chicken & egg problem.

> For instance, maybe get_extent gives you a single outer bounding box of of all the other ones, whereas other methods might just iterate over the children?

This approach to `_get_extent()` becomes tricky when mixing children with different dimensions (2D, 3D, 2D+time might have a `_get_extent()` we can't distinguish from the spatial 3D case). I can hack something that kinda works here, but it feels like a code smell to me.

> ... but I think each method is going to take some thought and there may not be universally applicable strategies.

Yeah, at this point I'm worried that most things will need some kind of special handling. (But maybe that's ok if we don't plan on adding any other layer class types?)

**Comment by sofroniewn** [comment]
> What should the .data property of a layergroup be?
>
> - Should it be None?
> - Should it be a list of all child .data results? A generator?
> - I can't really stack all the child .data results along a new array axis, because they all have very different shapes. Also, this is much more work to refresh/update on data changes for any of the layers (this criticism also applies to the list/generator idea above, too)
> - Should the .data attribute just not exist at all for layergroups?

Thanks for raising this @GenevieveBuckley. Maybe @royerloic can weigh in on this before the meeting today. I'll give it some thought too.

**Comment by royerloic** [comment]
I can't see a reason to have a .data field for layer groups. Perhaps that is the definition of a layergroup, a layer that does not have data but that groups layers that do, recursively.  It is fine to have the .data property in the layergroup base class return None. The question is what client code would need that and for what? My gut feeling is there is no client code that would need that field. 

Perhaps the real solution is to have a base class above the level of layers that does not have some fields (such as .data) and from which the layer base class derives. So this means that: 'a layer is a layer group that has data'  (and a few other things that groups don't need). This feels like the correct OO solution to this problem.  

About @GenevieveBuckley concerns on the 'chicken & egg problem'. I understand your concerns here, but I believe we can eventually solve the UI issues but only after we have a clear conceptualisation for the model. For example, for _get_extent() it is perfectly fine to return the bounding box with the 'bounding number of dimensions'. I am pretty confident that we will find a mathematically sound and elegant solution for all of these questions as we have done with opacity etc... We can go through each method together and discuss what should happen. I am not too worried. The model needs to make sense and be consistent, the connecting the wires with the UI will be easier...

**Comment by GenevieveBuckley** [comment]
Thank you for those helpful comments, @royerloic

**Comment by chrisroat** [comment]
The idea of groups sounds very interesting.  One additional thought regarding generic grouping would be to allow layers to belong to multiple groups.  For example, a layer could have multiple attributes, and visibility could be toggled by attribute.  In email client parlance, instead of a "folder" approach, it would be an "label" approach.

The label approach is more expressive and flexible -  but it would not be hierarchical, so it loses the benefits of propagating characteristics up or down a hierarchy.

**Comment by jwindhager** [comment]
Hi all, thankfully @sofroniewn pointed me to this issue & pull request in https://github.com/napari/napari/issues/2202. I'm not sure about the state of this, so I just would quickly like to share my thoughts here (also working on multiplexed imaging data), at the risk of stating the obvious/describing what's already implemented:
- A tree structure for images (hierarchically) contained in one file would be amazing to have, thanks for working on this!
- However, I would go a step further and propose to - in nomenclature and implementation - strictly differentiate between:
  - Files, where a file can be seen as a (potentially hierarchical) collection of images sharing a coordinate system
  - Images, i.e. TZCYX0 imaging data stacks, possibly with xarray-style axis labels (e.g. channel names)
  - Channels, i.e. user-friendly names (not indices!) to the C coordinate, possibly shared among images
  - Layers, i.e. shown (Z)YX0 2D planes/3D volumes corresponding to a TC combination
- IMHO, following up on @kevinyamauchi's comment and unlike in the original proposal, the tree structure should only cover file and image data items and channels shared among these items should be listed separately (not a tree structure!). This would allow the user to apply layer properties such as color, contrast, etc to channels (i.e., layers grouped by channel label). One could retain the Layers panel to still be able to apply properties to individual layers.

Having such a tree structure as well as a separation between files/images/channels/layers is essential for efficiently working with highly multiplexed imaging data with 50+ channels, with multiple images sharing the same (physical/machine) coordinate system. Adding full support for named channels would make napari immensely useful to the highly multiplexed imaging community! I described a use case for this in https://github.com/napari/napari/issues/2202 (Imaging Mass Cytometry).

I understand that the core of napari should be kept as generic as possible. To support this, one could e.g. employ a tree model in which file reader plugins can either populate a default node model, or provide their own custom one for specialized use cases.

In the long run, I think that having a separate channels list is crucial - is there a feature request already for this/should I open a new one separate from this discussion? If you feel like this is an issue, one could even think of making this as generic as possible, by replacing the concept of "channels" with a more general concept of "tree items grouped by one of their properties" (e.g. channel label), for which napari could then show list-like interfaces (e.g. in a tabs widget, one tab per group, or a combo box to select the grouping key).

Again, I'm not sure about the current state of this issue (sorry for being late to the conversation - I just started using/developing for napari and I love it!), but I think this proposal might help maturing napari to a fully fledged multi-channel analysis platform.

**Comment by jni** [comment]
Thanks for chiming in @jwindhager! Love the data and plugin you showed in #2202. Very cool stuff, and this kind of data that challenges how we think about layers is actually really important to making napari as generic as possible!

Until now, we've definitely conflated channels and layers, and I see the distinction now. :joy: One thing I would say though, is that channel is just one of many attributes that you might want to group layers by. For example, you might want to group all segmentation layers together and process them together, or group all mask layers. So I guess what I'm trying to say is that layergroups should not be strictly hierarchical but rather be flexible in some other ways — so that you could use it for channels but someone else could use it for a different purpose — say, treatment.

The UI/UX considerations are interesting to say the least though. :sweat_smile:

**Comment by jwindhager** [comment]
Thank you very much, @jni!

> One thing I would say though, is that channel is just one of many attributes that you might want to group layers by. For example, you might want to group all segmentation layers together and process them together, or group all mask layers. 

I fully support this! That's precisely what I meant with "tree items grouped by one of their properties", where I referred to images as "tree items" - apologies for the confusion!

> So I guess what I'm trying to say is that layergroups should not be strictly hierarchical but rather be flexible in some other ways — so that you could use it for channels but someone else could use it for a different purpose — say, treatment.

I would like to emphasize that I see the grouping and the tree structure inherent to some image formats as two orthogonal things: The tree structure is required for organizing images contained in a (hierarchical) file such as in the case of #2202, whereas the (flat) grouping of layers (that correspond to images in that tree structure, e.g. one per image and channel) is required to jointly apply settings to the same channel, condition, etc

> The UI/UX considerations are interesting to say the least though. sweat_smile

Certainly no easy task! :joy: Some naive ideas in https://github.com/napari/napari/issues/970#issuecomment-772433644

**Comment by sofroniewn** [comment]
> I would like to emphasize that I see the grouping and the tree structure inherent to some image formats as two orthogonal things: The tree structure is required for organizing images contained in a (hierarchical) file such as in the case of #2202, whereas the (flat) grouping of layers (that correspond to images in that tree structure, e.g. one per image and channel) is required to jointly apply settings to the same channel, condition, etc

While agreed they can be treated orthogonally, I think we were imaging a tree pattern where modifying a node attribute can modify all the properties of its children. What i realize now though is that you actually might want to "group" nodes that aren't heirachically related in the tree - i.e. you have three plates, each plate has red/ blue channel children and then you want to set the clims on all the blue channels in all three plates, but not the red channels. I don't think this would be easily supported with the API we were currently imagining, right @tlambert03? Or can you see a way?

**Comment by tlambert03** [comment]
nope, #970 and #1502 are literal groupings of layers, and not so much "binding" of certain layer attributes.  However, it's not so hard to imagine, a relatively simple API for binding the attributes of ~two~ multiple layers:

```python
from itertools import combinations

def link_layers(layers: List[Layer], attributes List[str]):
    for combo in combinations(layers, 2):
        for attribute in attributes:
            # e.g. layer_a.events.attribute.connect(_update_layer_b_attribute)
            connect_events_without_recursion(*combo)

```

**Comment by jwindhager** [comment]
> What i realize now though is that you actually might want to "group" nodes that aren't heirachically related in the tree - i.e. you have three plates, each plate has red/ blue channel children and then you want to set the clims on all the blue channels in all three plates, but not the red channels.

Precisely what I wanted to say! Sorry, it's really hard to explain these concepts. Maybe, with this realization in mind, https://github.com/napari/napari/issues/970#issuecomment-772433644 makes more sense now.

> nope, #970 and #1502 are literal groupings of layers, and not so much "binding" of certain layer attributes. However, it's not so hard to imagine, a relatively simple API for binding the attributes of two multiple layers:

In this case, should I open a new issue for discussing this separately (as proposed above)? I think having *both* a tree-structure for representing more complex file formats *and* an option to group layers would be key to multiplexed imaging use cases (meanwhile, for IMC data, the plugin mentioned in #2202 fulfills this requirement). Also happy to discuss this offline to avoid misunderstandings.

**Comment by jwindhager** [comment]
As I realized that my request is somewhat different to this issue, I created a new feature request for this in #2229. Apologies for hijacking this conversion with partially unrelated input!

**Comment by isabela-pf** [comment]
Heads up that I'm planning to look into layer groups work from a user experience and design perspective! I'm currently planning to add updates to this issue, and I'll let you know here if that changes.

**Comment by matham** [comment]
I'm thinking this could be useful for my use-case with segmentation.

I have a 3d volume with a bunch of cells. I've segmented the cells and then like to highlight a mask or intensity values of the segmentation for selected cells. For one cell it's easy: create an image layer with the volume being the size of the segmented cell, fill in the the voxels as a mask/intensity, set the translation to the corner of the cell and it overlays the cell perfectly.

However, if I want to look at 10 or 100 cells, it would create 10 or 100 layers, which is annoying to navigate. Obviously I can create a large array and fill in all the cells, but that may be a bit much memory-wise. So having some kind of nested layering situation would make it easier navigationally, as long as they can have their individual `translate` values.
